add_source_file

```diff
@@ -0,0 +1,238 @@
+# should be the wesnoth root directory
+rootdir = pathlib.Path(inspect.getsourcefile(lambda:0))
+if not rootdir.joinpath("projectfiles").exists():
+    rootdir = pathlib.Path()
+if not rootdir.joinpath("projectfiles").exists():
+    raise Exception("Could not find project file directory")
+
+# the names of the targets in the Xcode project
+xcode_target_translations = {
+    "wesnoth": "The Battle for Wesnoth",
+    "wesnothd": "wesnothd",
+    "campaignd": "campaignd",
+    "lua": "liblua",
+    "tests": "unit_tests",
+}
+
+# the names of the targets in source_lists
+source_list_target_translations = {
+    "wesnoth": "wesnoth",
+    "wesnothd": "wesnothd",
+    "campaignd": "campaignd",
+    "lua": "lua",
+    "tests": "boost_unit_tests",
+}
+
+# the names of the targets in Code::Blocks
+code_blocks_target_translations = {
+    "wesnoth": "wesnoth",
+    "wesnothd": "wesnothd",
+    "campaignd": "campaignd",
+    "lua": "liblua",
+    "tests": "tests",
+}
+
+#=======#
+# XCode #
+#=======#
+
+def add_to_xcode(filename, targets):
+    """Add the given file to the specified targets.
+    """
+    projectfile = rootdir.joinpath(
+        "projectfiles",
+        "Xcode",
+        "The Battle for Wesnoth.xcodeproj",
+        "project.pbxproj",
+    )
+    
+    project = pbxproj.XcodeProject.load(projectfile)
+    
+    translated_targets = [xcode_target_translations[t] for t in targets]
+    print(" xcode targets:", translated_targets)
+    
+    for tname in translated_targets:
+        if not project.get_target_by_name(tname):
+            raise Exception(
+                f"Could not find target '{tname}' in Xcode project file")
+    
+    # groups are organized by directory structure under "src"
+    src_groups = project.get_groups_by_name("src")
+    if len(src_groups) != 1:
+        raise Exception("problem finding 'src' group in xcode project")
+    src_group = src_groups[0]
+    parent_group = src_group
+    for d in filename.parts[:-1]:
+        found_groups = project.get_groups_by_name(d, parent=parent_group)
+        if len(found_groups) != 1:
+            groupname = parent_group.get_name()
+            raise Exception(f"problem finding '{d}' group in '{groupname}'")
+        parent_group = found_groups[0]
+    
+    # if the group already has an entry with the same filename, loudly skip.
+    # note: this doesn't allow adding to targets one at a time.
+    # a new file should be added to all targets at once...
+    # or maybe targets could be checked somehow,
+    # or maybe the file could simply be completely removed and readded.
+    if project.get_files_by_name(filename.name, parent=parent_group):
+        print("  '{filename}' already found in Xcode project, skipping")
+        return
+    
+    # force is True here because otherwise a duplicate filename in
+    # a different place will block addition of the new file.
+    # the rest is just to match existing project file structure.
+    project.add_file(filename,
```

Comment:
Does this keep the files in the group sorted? I always add new files at the alphabetical location.

Reply:
It seems to append to the end of the group. I'm not sure if there's a way to change this behaviour.